### PR TITLE
Reclaim paragraph memory more aggressively

### PR DESCRIPTION
**Pull request by yjbanov**
Put `CkParagraph` objects into a separate synchronous cache, which reclaims memory synchronously. This significantly reduces paragraph memory usage for very long `ListView`s and other cases where lots of paragraphs are laid out but not rendered. It does not fully cover the use-case where we both lay out and render tons of paragraphs (it fixes the memory issue but not the performance issue). For that we need https://github.com/flutter/flutter/issues/81224.

Some benchmark numbers:

`text_canvas_kit_color_grid`: before 420MB, after 153MB
Layout and render 10000 paragraphs: before 1GB, after 200MB
[sample 1, 60 seconds](https://github.com/flutter/flutter/issues/66614#issue-708581477) (by @slavap): before 600MB, after 168MB
[sample 2, 100 tab switches](https://github.com/flutter/flutter/issues/66614#issuecomment-844733715) (by @asjqkkkk): before 422MB, after 253MB

Fixes https://github.com/flutter/flutter/issues/66614

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/text.dart [hunk 1]
```diff
@@ -553,74 +563,138 @@ class CkParagraph extends ManagedSkiaObject<SkParagraph>
...
+  static SynchronousSkiaObjectCache _paragraphCache = SynchronousSkiaObjectCache(500);
+
+  /// Marks this paragraph as eligible for GC.
+  ///
+  /// Puts this paragraph in a [SynchronousSkiaObjectCache], which will delete it
+  /// if there memory pressure to do so. This protects our memory usage from
+  /// blowing up if within a single frame the framework needs to layout a lot of
+  /// paragraphs. One common use-case is `ListView.builder`, which needs to layout
+  /// more of its content than it actually renders to compute the scroll position.
+  void release() {
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/canvaskit/text.dart:645]: I think this name is a little confusing. `release` makes me think we are telling the cache to delete this paragraph ASAP. I think something like `addToCache` or `markUsed` would make more sense

**Review by hterkelsen — COMMENTED** [review]
This approach seems good to me

File: lib/web_ui/lib/src/engine/canvaskit/text.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvaskit/text.dart]: Good point. Done.

**Review by hterkelsen — APPROVED** [review]
LGTM

**Pull request merged**